C4.md

```diff
@@ -0,0 +1,131 @@
+2.  To relieve dependencies on key individuals by separating different skill sets so that there is a larger pool of competence in any required domain;
+3.  To allow the project to develop faster and more accurately, by increasing the diversity of the decision making process;
+4.  To support the natural life cycle of project versions from experimental through to stable, by allowing safe experimentation, rapid failure, and isolation of stable code;
+5.  To reduce the internal complexity of project repositories, thus making it easier for Contributors to participate and reducing the scope for error;
+6.  To enforce collective ownership of the project, which increases economic incentive to Contributors and reduces the risk of hijack by hostile entities.
+
+## 2. Design
+
+### 2.1. Preliminaries
+
+1.  The project SHALL use the git distributed revision control system.
+2.  The project SHALL be hosted on github.com or equivalent, herein called the “Platform”.
+3.  The project SHALL use the Platform issue tracker.
+4.  The project SHOULD have clearly documented guidelines for code style.
+5.  A “Contributor” is a person who wishes to provide a patch, being a set of commits that solve some clearly identified problem.
+6.  A “Maintainer” is a person who merges patches to the project. Maintainers are not developers; their job is to enforce process.
+7.  An “Administrator” is a person who manages the set of Maintainers. They are Maintainers themselves.
+8.  A “Founder” is a person who created the project and/or is moral owner in some sense.
+9.  Contributors SHALL NOT have commit access to the repository unless they are also Maintainers.
+10. Maintainers SHALL have commit access to the repository.
+11. Administrators SHALL have administration rights on the Platform.
+12. Everyone, without distinction or discrimination, SHALL have an equal right to become a Contributor under the terms of this contract.
+
+### 2.2. Licensing and Ownership
+
+1. The project SHALL use a share-alike license such as the MPLv2, or a GPLv3 variant thereof (GPL, LGPL, AGPL).
+2. All contributions to the project source code (“patches”) SHALL use the same license as the project.
+3. All patches are owned by their authors. There SHALL NOT be any copyright assignment process.
+4. Each Contributor SHALL be responsible for identifying themselves in the project Contributor list.
+
+### 2.3. Patch Requirements
+
+1.  Maintainers and Contributors MUST have a Platform account and SHOULD use their real names or a well-known alias.
+2.  A patch SHOULD be a minimal and accurate answer to exactly one identified and agreed problem.
+3.  A patch MUST adhere to the code style guidelines of the project if these are defined.
+4.  A patch MUST adhere to the “Evolution of Public Contracts” guidelines defined below.
+5.  A patch SHALL NOT include non-trivial code from other projects unless the Contributor is the original author of that code.
```

Comment:
I'm a little unsure just by reading this: does this apply to the inclusion of new libraries not written by the contributor? Because if so, I feel like it might be a bit over-restrictive.

We *should* ensure that any libraries introduced don't have negative side effects or licensing issues, but that's a whole other topic.

Reply:
A "patch" is normally a single problem that need to be fixed. It's not really about libraries, it's about fixing this problem using a patch. The rationale behind this argument is to solve the patch by not introducing non-trivial code from others (assuming it's not causing license issues) but if a patch is needed that is non-trivial the code _could_ only come from the contributor itself. Ideally **all code** should be trivial of course and easy to understand.

Reply:
Makes total sense to me; I think I was just over-thinking it and assumed that "patch" was also overloaded to mean "feature implementation", but from the rest of the discussion I think it's truly more limited along the lines of a bugfix.